### PR TITLE
[YUNIKORN-1709] Add event streaming logic

### PR DESCRIPTION
**Pull request by pbacsko**
### What is this PR for?
Core implementation of event streaming. Clients are updated real time on the REST interface. A persistent HTTP connection is maintained for them.

Consumers can request a new stream and receive core scheduler events immediately on a read-only channel. Slow consumers are detected by checking the number of elements in the channel buffer. If the channel buffer on the local side is full, we remove the consumer.

Explicit closure of the event stream is possible and consumers are expected to do it.

### What type of PR is it?
* [ ] - Bug Fix
* [ ] - Improvement
* [x] - Feature
* [ ] - Documentation
* [ ] - Hot Fix
* [ ] - Refactoring

### Todos
* [ ] - Task

### What is the Jira issue?
https://issues.apache.org/jira/browse/YUNIKORN-1709

### Questions:
* [ ] - The licenses files need update.
* [ ] - There is breaking changes for older versions.
* [ ] - It needs documentation.

### CODE REVIEW

File: pkg/events/event_ringbuffer.go [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+//
+// Retrieving the records can be achieved with GetLatestEntriesCount and GetLatestEntries. Since these do not
+// remove the elements, they are not regular dequeue operations either.
+//
+// Entries have a maximum lifespan defined in nanoseconds. Cleanup of expired objects occurs when a call to
+// RemoveExpiredEntries is made.
+//
+// The buffer does not use locking, proper synchronization is delegated to the client.
+type EventRingBuffer struct {
+	events        []*si.EventRecord
```
Comment (pbacsko) [inline pkg/events/event_ringbuffer.go:44]: IMO we need to switch to `[]si.EventRecord` if we want to store a lot of elements. We want to avoid GC interactions, so using a value type here is very much reasonable.

In fact, I'm considering a specia, internal type here instead of `EventRecord` with interned strings. There will be a lot of string duplications in the events (especially IDs). 
This library looks promising: https://github.com/philpearl/intern

File: pkg/events/event_streaming.go [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+}
+
+// PublishEvent publishes an event to event stream consumers. Events are first stored in a ring buffer,
+// then all clients are notified. Clients whose buffer is full are considered slow and removed.
+func (e *EventStreaming) PublishEvent(event *si.EventRecord) {
+	e.Lock()
+	defer e.Unlock()
+
+	e.buffer.Add(event)
+	for consumer, details := range e.eventStreams {
```
Comment (pbacsko) [inline pkg/events/event_streaming.go:65]: We can add another level of channel communication here, so updating clients will occur on a separate goroutine. So this method can look like:

```
func (e *EventStreaming) PublishEvent(event *si.EventRecord) {
	e.Lock()
	defer e.Unlock()
	e.buffer.Add(event)
        e.events <- event
}
```

So the event flow will be:
incoming event --> `e.events` --> `consumerDetails.local` --> `consumerDetails.consumer`.

It's a matter of taste, this method is already fast and not affected by slow clients.

**Comment by pbacsko** [comment]
Note: we can hold off merging until 1.4.0 branching is ready. This is intended for Yunikorn 1.5.0.

**Comment by pbacsko** [comment]
TODO:
- http status codes on return & error object
- idle timeout
- coverage

**Comment by pbacsko** [comment]
Note: any timeout for the HTTP server can break the streaming (WriteTimeout for 1.1, ReadTimeout and IdleTimeout for 2.0). We either start a new server instance for the streaming or explicitly comment (or even test)  that timeouts must not be used.

File: pkg/webservice/handlers.go [hunk 3]
```diff
@@ -918,3 +918,61 @@ func getEvents(w http.ResponseWriter, r *http.Request) {
 		buildJSONErrorResponse(w, err.Error(), http.StatusInternalServerError)
 	}
 }
+
+func getStream(w http.ResponseWriter, r *http.Request) {
```
Comment (wilfred-s) [inline pkg/webservice/handlers.go:922]: We should limit the number of streams that we allow to be open at the same time here. Even if it is some "high" number but we need to be able to turn down new requests as a safety mechanism.

File: pkg/webservice/handlers.go [hunk 4]
```diff
@@ -918,3 +918,61 @@ func getEvents(w http.ResponseWriter, r *http.Request) {
...
+			return
+		}
+	}
+
+	enc := json.NewEncoder(w)
+	stream := eventSystem.CreateEventStream(r.Host, count)
+
+	// Reading events in an infinite loop until either the client disconnects or Yunikorn closes the channel.
+	// This results in a persistent HTTP connection where the message body is never closed.
+	// We don't use timeouts and since HTTP 1.1 clients are expected to handle persistent connections by default.
```
Comment (wilfred-s) [inline pkg/webservice/handlers.go:951]: See comment earlier, we should use the ResponseController with a sliding  write deadline.

File: pkg/scheduler/objects/common_test.go
Comment (wilfred-s) [inline pkg/scheduler/objects/common_test.go:1]: not sure why these changes are needed.

Comment (pbacsko) [reply pkg/scheduler/objects/common_test.go:1]: Interfaces changes in EventSystem (two new methods).

File: pkg/webservice/handlers.go (hunk 3 above)
Comment (pbacsko) [reply pkg/webservice/handlers.go:922]: Yes, there is a follow up ticket for that: https://issues.apache.org/jira/browse/YUNIKORN-2147

File: pkg/webservice/webservice.go [hunk 5]
```diff
@@ -63,12 +64,13 @@ func loggingHandler(inner http.Handler, name string) http.HandlerFunc {
 // TODO we need the port to be configurable
 func (m *WebService) StartWebApp() {
 	router := newRouter()
+	// Important: do not use ReadTimeout, WriteTimeout or IdleTimeout because those can break the event streaming
```
[the remark this answers is not in the document]
Comment (pbacsko) [reply pkg/webservice/webservice.go:67]: `ReadHeaderTimeout` is fine, it's the rest what can cause issues. Will check out ResponseController, looks like it's relevant to us.

File: pkg/webservice/handlers.go (hunk 4 above)
Comment (pbacsko) [reply pkg/webservice/handlers.go]: Change added. New code uses ResponseController.setWriteDeadline() before each send.

File: pkg/webservice/webservice.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (pbacsko) [reply pkg/webservice/webservice.go:67]: Write timeout added

File: pkg/events/event_streaming.go [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+// Events are sent to the "local" channel from where it is forwarded to the "consumer" channel.
+//
+// If "local" is full, it means that the consumer side has not processed the events at an appropriate pace.
+// Such a consumer is removed and the related channels are closed.
+func (e *EventStreaming) PublishEvent(event *si.EventRecord) {
+	e.Lock()
+	defer e.Unlock()
+
+	for consumer, details := range e.eventStreams {
+		if len(details.local) == defaultChannelBufSize {
```
Comment (zhuqi-lucas) [inline pkg/events/event_streaming.go:67]: Minor question, if we can set custom defaultChannelBufSize or dynamic range for the size, here 100 seems small for some cases.

File: pkg/events/event_streaming.go [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+// The streaming logic uses bridging to ensure proper ordering of existing and new events.
+// Events are sent to the "local" channel from where it is forwarded to the "consumer" channel.
+//
+// If "local" is full, it means that the consumer side has not processed the events at an appropriate pace.
+// Such a consumer is removed and the related channels are closed.
+func (e *EventStreaming) PublishEvent(event *si.EventRecord) {
+	e.Lock()
+	defer e.Unlock()
+
+	for consumer, details := range e.eventStreams {
```
Comment (zhuqi-lucas) [inline pkg/events/event_streaming.go:66]: @wilfred-s @pbacsko I think the customer here is a good place for us to set a internal customer, the new trackingService, we can at least use it to streaming resourcce and aggregated tracking events.

File: pkg/events/event_streaming.go (hunk 6 above)
Comment (pbacsko) [reply pkg/events/event_streaming.go:67]: Yes 100 might be a bit too low. 1000 sounds like a reasonable value.

File: pkg/events/event_streaming.go (hunk 7 above)
Comment (pbacsko) [reply pkg/events/event_streaming.go:66]: If we decided to this internally, we might as well just run the aggregation on a separate goroutine. 

Wilfred's comments:
"I think we can simply use the existing events and push the whole aggregation out of YuniKorn. The only thing that seems to be missing is the node instance type. If that is available just listening to the app and node events would allow creating the summary outside of the scheduler.
[...]
hat is the best option. A scheduler should schedule. It is not for providing statistics."

To me it sounds like a separate application.

Comment (zhuqi-lucas) [reply pkg/events/event_streaming.go:66]: Yeah, we can have a separate service, but we can let the service as a cusomer here, so we don't need to add another publisher, just one publisher for the streaming, and if we have several  publishers, it seems will cost more overhead.

Comment (pbacsko) [reply pkg/events/event_streaming.go:66]: Yes, a separate service will use the REST API (either batch or streaming, whatever is more convenient), just like everyone else. If this is what you mean, I agree with it.

File: pkg/events/event_streaming.go (hunk 6 above)
Comment (zhuqi-lucas) [reply pkg/events/event_streaming.go:67]: Agree , 1000 seems more reasonable

Comment (pbacsko) [reply pkg/events/event_streaming.go:67]: Changed

File: pkg/events/event_streaming.go [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+	}
+}
+
+// CreateEventStream sets up event streaming for a consumer. The returned EventStream object
+// contains a channel that can be used for reading.
+//
+// When a consumer is finished, it must call RemoveEventStream to free up resources.
+//
+// Consumers have an arbitrary name for logging purposes. The "count" parameter defines the number
+// of maximum historical events from the ring buffer.
```
Comment (zhuqi-lucas) [inline pkg/events/event_streaming.go:83]: Nit:

Add more comments about the count, when we want to get all the historical events from the ring buffer, we can set count to 0

File: pkg/events/event_streaming.go [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+				consumer <- event
+			}
+		}
+	}(consumer, local, stop)
+
+	log.Log(log.Events).Info("Created event stream", zap.String("consumer name", name))
+	return stream
+}
+
+func (e *EventStreaming) createEventStreamInternal(stream *EventStream,
```
Comment (zhuqi-lucas) [inline pkg/events/event_streaming.go:129]: Nit
The function name is 
createEventStreamInternal 

But we return the history events for this function, maybe we can get history events after we call this method.

File: pkg/events/event_streaming.go [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+func (e *EventStreaming) RemoveEventStream(consumer *EventStream) {
+	e.Lock()
+	defer e.Unlock()
+
+	e.removeEventStream(consumer)
+}
+
+func (e *EventStreaming) removeEventStream(consumer *EventStream) {
+	if details, ok := e.eventStreams[consumer]; ok {
+		log.Log(log.Events).Info("Removing event stream consumer", zap.String("name", details.name),
```
Comment (zhuqi-lucas) [inline pkg/events/event_streaming.go:160]: Nit 
Is it possible, we can logging the remove reason, such as:

"Connection closed for event stream client"
"Cannot set write deadline"
"Marshalling error"

**Review by zhuqi-lucas — APPROVED** [review]
LGTM Left final minor comments!

File: pkg/events/event_streaming.go (hunk 10 above)
Comment (pbacsko) [reply pkg/events/event_streaming.go:159]: The external reason why it was closed is logged at the appropriate place in the handler. Here, the code doesn't have to know about it.

File: pkg/events/event_streaming.go (hunk 8 above)
Comment (pbacsko) [reply pkg/events/event_streaming.go]: Actually "0" means no events. But extended the comment.

File: pkg/events/event_streaming.go [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+}
+
+// CreateEventStream sets up event streaming for a consumer. The returned EventStream object
+// contains a channel that can be used for reading.
+//
+// When a consumer is finished, it must call RemoveEventStream to free up resources.
+//
+// Consumers have an arbitrary name for logging purposes. The "count" parameter defines the number
+// of maximum historical events from the ring buffer. "0" is a valid value and means no past events.
+func (e *EventStreaming) CreateEventStream(name string, count uint64) *EventStream {
```
Comment (zhuqi-lucas) [inline pkg/events/event_streaming.go:84]: Hi @pbacsko , i am thinking if we can have an option can include all history events without setting any count, thanks!

Comment (zhuqi-lucas) [reply pkg/events/event_streaming.go:84]: Sure, it's a solution, agree!

File: pkg/events/event_system.go [hunk 12]
```diff
@@ -94,21 +141,25 @@ func (ec *EventSystemImpl) GetRingBufferCapacity() uint64 {
...
+		streaming:     NewEventStreaming(buffer),
 	}
 }
 
+// StartService starts the event processing in the background. See the interface for details.
 func (ec *EventSystemImpl) StartService() {
 	ec.StartServiceWithPublisher(true)
 }
 
+// VisibleForTesting
```
Comment (wilfred-s) [inline pkg/events/event_system.go:162]: fix comment to use standard go layout as it is an.exported function now.

File: pkg/events/event_system.go [hunk 13]
```diff
@@ -191,11 +245,21 @@ func (ec *EventSystemImpl) isRestartNeeded() bool {
 	return ec.readIsTrackingEnabled() != ec.trackingEnabled
 }
 
+// Restart restarts the event system, used during config update.
 func (ec *EventSystemImpl) Restart() {
 	ec.Stop()
 	ec.StartServiceWithPublisher(true)
 }
 
+// VisibleForTesting
```
Comment (wilfred-s) [inline pkg/events/event_system.go:254]: fix comment to use standard go layout as it is an.exported function now.

File: pkg/events/event_ringbuffer.go [hunk 14]
```diff
@@ -70,14 +70,38 @@ func (e *eventRingBuffer) Add(event *si.EventRecord) {
 	e.id++
 }
 
-// GetEventsFromID returns "count" number of event records from "id" if possible. The id can be determined from
+func (e *eventRingBuffer) GetRecentEvents(count uint64) []*si.EventRecord {
```
Comment (wilfred-s) [inline pkg/events/event_ringbuffer.go:73]: add comment use standard go layout as it is an.exported function

**Review by wilfred-s — APPROVED** [review]
minor changes to comments on exported functions can be fixed during the commit also

File: pkg/events/event_streaming.go [hunk 15]
```diff
@@ -0,0 +1,179 @@
...
+	log.Log(log.Events).Info("Created event stream", zap.String("consumer name", name))
+	return stream
+}
+
+func (e *EventStreaming) createEventStreamInternal(stream *EventStream,
+	local chan *si.EventRecord,
+	consumer chan *si.EventRecord,
+	stop chan struct{},
+	name string,
+	count uint64) {
```
[the remark this answers is not in the document]
Comment (wilfred-s) [reply pkg/events/event_streaming.go:135]: Need a follow up jira to clean that up @pbacsko

[the remark this answers is not in the document]
Comment (pbacsko) [reply pkg/events/event_streaming.go:135]: Correct. I actually have a list of necessary small code changes which are unrelated to each other, I'm adding this to the list.
